File: ts/src/unit-registry.ts
/**
 * Registry for measurement units, enabling lookup by ID.
 */

import { MeasurementUnit } from './measurement-unit';

export class UnitRegistry {
  private readonly byId = new Map<string, MeasurementUnit>();

  /** Adds a unit to the registry. Throws if ID is already registered. */
  register(unit: MeasurementUnit): void {
    if (this.byId.has(unit.id)) {
      throw new Error(`unit with id '${unit.id}' is already registered`);
    }
    this.byId.set(unit.id, unit);
  }

  /** Looks up a unit by ID. Throws if not found. */
  resolve(id: string): MeasurementUnit {
    const unit = this.byId.get(id);
    if (unit === undefined) {
      throw new Error(`unknown unit id: '${id}'`);
    }
    return unit;
  }

  /** Returns a registry pre-populated with Number, Ratio, and Disparity units. */
  static standard(): UnitRegistry {
    const r = new UnitRegistry();
    r.register(MeasurementUnit.NUMBER);
    r.register(MeasurementUnit.RATIO);
    r.register(MeasurementUnit.DISPARITY);
    return r;
  }
}

File: go/unit_registry.go
package pragmastat

import "fmt"

// UnitRegistry stores measurement units and enables lookup by ID.
type UnitRegistry struct {
	byID map[string]*MeasurementUnit
}

// NewUnitRegistry creates an empty registry.
func NewUnitRegistry() *UnitRegistry {
	return &UnitRegistry{
		byID: make(map[string]*MeasurementUnit),
	}
}

// Register adds a unit to the registry.
func (r *UnitRegistry) Register(unit *MeasurementUnit) error {
	if _, exists := r.byID[unit.ID]; exists {
		return fmt.Errorf("unit with id '%s' is already registered", unit.ID)
	}
	r.byID[unit.ID] = unit
	return nil
}

// Resolve looks up a unit by ID.
func (r *UnitRegistry) Resolve(id string) (*MeasurementUnit, error) {
	if unit, ok := r.byID[id]; ok {
		return unit, nil
	}
	return nil, fmt.Errorf("unknown unit id: '%s'", id)
}

// StandardRegistry returns a registry pre-populated with Number, Ratio, and Disparity units.
func StandardRegistry() *UnitRegistry {
	r := NewUnitRegistry()
	_ = r.Register(NumberUnit)
	_ = r.Register(RatioUnit)
	_ = r.Register(DisparityUnit)
	return r
}
